new(HierarchyPicker): Add new picker component

williaster opened the pull request:
to: @milesj @stefhatcher @hayes @alecklandgraf
cc: @elibrumbaugh @jpdanks 

## Motivation and Context

Implements @jpdanks 's 2.0 version of the picker.

## Description

This PR adds a `HierarchyPicker` `"2.0"` to `@airbnb/lunar`, and is derived mostly from the old `HierarchyPicker` `"1.0"` component (which was removed in the open-source move).

For easier code review, the component architecture / logic basically looks like the following (this is the same vs 1.0):
```jsx
<>
  <FauxSelect>
  <Portal>
    <Autocomplete />
    <Hierarchy />
  </Portal>
</>
```

The follow changes were made vs the `"1.0"` component:

#### New
- adds `verticallyAlign` prop which will result in vertical alignment (vs vertical offset) of nested hierarchy lists. DOM-wise, this results in **_sibling_** `ul`s rather than _**nested**_ `ul`s and required updates to the hierarchy focus traversal logic
- adds more hooks to control the sizing of the `HierarchyPicker` pieces:
  - `hierarchyWidth` sets the width of a given level of hierarchy 
  - `hierarchyMaxHeight` sets the overflow height of a given hierarchy level. the height of the overall hierarchy picker is the `max` of any visible nested menu. 
  - `searchWidth` sets the width of the hierarchy search input
  - `searchMaxHeight` sets the max height of the search result list
- adds `renderItem: (item, selected, focused) => ReactNode` to override rendering of a hierarchy list item
- adds general support for sub-section headings/organization within hierarchy menu by specifying a `section` string in the `items` definition (example in the readme)

#### Breaking (vs 1.0)
- previously the component took a `suggestions` list, which provided a limited API to organize the menu. suggestions are **no longer supported**, and generic support for menu sections was added.

 #### Refactor
- refactors hierarchy search to leverage the `Autocomplete` component 

## Screenshots

**Vertically offset**
<kbd>
  <img src="https://user-images.githubusercontent.com/4496521/56632857-ac6f5c80-6610-11e9-81a2-74a4ad495648.png" width="600" />
</kbd>

**Vertically aligned, with subheadings**
<kbd>
  <img src="https://user-images.githubusercontent.com/4496521/56632887-c741d100-6610-11e9-82f6-775307c3c099.png" width="600" />
</kbd>

**Keyboard nav demo**
<kbd>
  <img src="https://user-images.githubusercontent.com/4496521/56633427-ef323400-6612-11e9-9ee5-5dfd62fdc5c0.gif" width="600" />
</kbd>

**Custom hierarchy width**
<kbd>
  <img width="600" src="https://user-images.githubusercontent.com/4496521/56633057-6f579a00-6611-11e9-9b34-038690610e93.png" />
</kbd>

**Custom search dimensions**
<kbd>
  <img width="600" src="https://user-images.githubusercontent.com/4496521/56633148-d412f480-6611-11e9-9f13-0c9d3fc6e131.png" />
</kbd>

## Testing

- [x] functional
- [x] add unit tests (WIP)

## Checklist

- My code follows the style guide of this project.
- I have updated or added documentation accordingly.
- I have read the CONTRIBUTING document.

milesj commented on packages/core/package-lock.json:
```diff
@@ -1,6 +1,6 @@
 {
```
These diffs look wrong.

williaster replied:
I think this looks okay now. I'm not sure what had the `fuse.js` dep before tho 🤔 `npm ls fuse.js` only shows this package.

milesj commented on packages/core/src/components/HierarchyPicker.story.tsx:
```diff
@@ -0,0 +1,304 @@
+            name: 'General',
+          },
+          {
+            name: 'Suspension',
```
Can we update all the copy for these items to be a bit more... nice? And in English. Can even just be something like "Item #1".

milesj commented on packages/core/src/components/HierarchyPicker/Hierarchy/HierarchyList.tsx:
```diff
@@ -0,0 +1,287 @@
+                      focused={isFocused}
+                    />
+
+                    {!verticallyAlign && isFocused && item.items ? (
```
`item.items.length > 0`

williaster replied:
the `1.0` recursive logic handles an empty array but I agree this is better 👍

milesj commented on packages/core/src/components/HierarchyPicker/Hierarchy/HierarchyList.tsx:
```diff
@@ -0,0 +1,287 @@
+          </ul>
+        </div>
+
+        {verticallyAlign && focusedItem && focusedItem!.items ? (
```
`focusedItem!.items.length > 0`

milesj commented on packages/core/src/components/HierarchyPicker/Hierarchy/helpers.ts:
```diff
@@ -0,0 +1,24 @@
+import { ItemShape } from '../types';
+
+export function allChildrenReadonly(parent: ItemShape): Boolean {
```
`boolean` not `Boolean`. Be sure to apply this everywhere as I've seen it a few times.

williaster replied:
not added by me but 👍

milesj commented on packages/core/src/components/HierarchyPicker/Picker/index.tsx:
```diff
@@ -0,0 +1,173 @@
+import React from 'react';
+
```
Remove these lines between imports.

milesj commented on packages/core/src/components/HierarchyPicker/Search/Highlight.tsx:
```diff
@@ -0,0 +1,86 @@
+          <span {...css(styles.highlight)} key={i}>
+            <span
+              style={{
+                opacity: sacc.trim().toLowerCase() === word ? 1 : 0.33,
```
This can be done with styles since have now have a range of 7 shades.

milesj commented on packages/core/src/components/HierarchyPicker/Search/SearchResult.tsx:
```diff
@@ -0,0 +1,65 @@
+    const [labelMatch = null] = mbk.label || [];
+    const [descMatch = null] = mbk.description || [];
+    const [keywMatch = null] = mbk.keywords || [];
+    const [longest] = query!.split(' ').sort((a, b) => b.length - a.length);
```
`split(/\s{1,}/)`?

milesj commented on packages/core/src/components/HierarchyPicker/Search/fuseLoader.ts:
```diff
@@ -0,0 +1,10 @@
+import Fuse from 'fuse.js';
+
+export default async function fuseLoader(
```
We should get rid of this stuff. We can just import fuse directly instead of `import()`ing it.

williaster replied:
is there any downside to this? there's a lot of code that's based on this assumption sprinkled throughout components and tests.

milesj commented on packages/core/src/components/HierarchyPicker/Search/fuseLoader.mock.js:
```diff
@@ -0,0 +1,7 @@
+// This mock exists because the normal fuseLoader uses `import()` which jest doesn't understand.
```
Remove this.

milesj commented on packages/core/src/components/HierarchyPicker/Search/index.tsx:
```diff
@@ -0,0 +1,178 @@
+  maxHeight?: number;
+};
+
+const defaultFuseOptions: Fuse.FuseOptions<any> = {
```
Avoid using `any`. We should type the structure correctly.

williaster replied:
didn't add this but will update

milesj commented on packages/core/src/components/HierarchyPicker/Search/index.tsx:
```diff
@@ -0,0 +1,178 @@
+    query: '',
+  };
+
+  fusePromise?: Promise<Fuse<SearchItemResult>>;
```
Can ditch this since we shouldn't `import()`.

milesj commented on packages/core/src/components/HierarchyPicker/index.tsx:
```diff
@@ -0,0 +1,262 @@
+import React from 'react';
+import { FuseOptions } from 'fuse.js';
+import IconCaretDown from '@airbnb/lunar-icons/lib/interface/IconCaretDown';
+
```
Remove newline.

milesj commented on packages/core/src/components/HierarchyPicker/index.tsx:
```diff
@@ -0,0 +1,262 @@
+  /** A function to format the display of choice. */
+  formatter?: (chosen: TreePath, labeler: Labeler) => string;
+  /** Fuse.js search options to override. */
+  fuseOptions?: FuseOptions<any>;
```
Avoid `any`.

milesj commented on packages/core/src/components/HierarchyPicker/types.ts:
```diff
@@ -0,0 +1,66 @@
+  matches: FuseMatch[];
+};
+
+export enum TopicOriginKey {
```
Use a string union instead of an enum.

milesj commented on packages/core/src/components/HierarchyPicker/index.tsx:
```diff
@@ -0,0 +1,262 @@
+}
+
+export default withStyles(({ color, unit, ui, pattern }) => ({
+  selectlike: {
```
We should probably inherit styles from `buildInputStyles`, instead of doing this custom.

williaster commented:
@milesj this should be good to go. I'll redo the `package-lock`s tho, root shouldn't have changed.

williaster commented:
can't fix the danger issue 🤷‍♀️

milesj commented:
@williaster I've been testing it in storybook, my feedback:

- We should avoid horizontal scrolling, it should just wrap. Noticed this a lot in the search results and the overflow tooltips.
<img width="426" alt="Screen Shot 2019-04-26 at 10 27 50 AM" src="https://user-images.githubusercontent.com/143744/56825438-f709f800-680d-11e9-80dd-af7836601a5e.png">
<img width="487" alt="Screen Shot 2019-04-26 at 10 29 07 AM" src="https://user-images.githubusercontent.com/143744/56825497-2456a600-680e-11e9-9111-11d7c50c1be9.png">

- Colleen and I both agree that the search input at the top is a bit... ugly? Padding around the input might be best, even if it's just 4 or 8px.
<img width="255" alt="Screen Shot 2019-04-26 at 10 32 20 AM" src="https://user-images.githubusercontent.com/143744/56825662-a941bf80-680e-11e9-8e58-e8028441cbaa.png">

- When searching, if you hover over an item, you'll see 2 flashes of gray backgrounds happening. Not sure why.

- For the selected path, can we use the caret icon instead of `>`?
<img width="258" alt="Screen Shot 2019-04-26 at 10 34 35 AM" src="https://user-images.githubusercontent.com/143744/56825752-f02fb500-680e-11e9-8fda-cfbf8a6aad9e.png">

williaster commented:
@milesj 

- Updated overflow with `word-break: 'break-word'` 
- added padding around `Autocomplete`, agree it looks a lot better 👌 (chose `8px` because it better aligns with the menu items)
- I agree `chevron`s look better but we can't easily update from `">"` because there's a lot of logic (including in `fuse`) that expects `formatter` to return a `string`.
- I'm not entirely sure what you're referring to with the double gray upon search. there's a debounced `focus` for the menu items, maybe that's something to do with it?


![image](https://user-images.githubusercontent.com/4496521/56830943-a1891780-681c-11e9-8793-41a1939e2aea.png)

![image](https://user-images.githubusercontent.com/4496521/56831085-12303400-681d-11e9-9a3c-a2d244738824.png)

williaster commented:
Closing in favor of #50
